src/audio/codec_adapter/codec/generic.c

```diff
@@ -229,17 +229,17 @@ int codec_prepare(struct comp_dev *dev)
 		goto end;
 	}
 
-	codec->s_cfg.avail = false;
-	codec->r_cfg.avail = false;
-	codec->r_cfg.data = NULL;
-
 	/* After prepare is done we no longer need runtime configuration
 	 * as it has been applied during the procedure - it is safe to
 	 * free it.
 	 */
 	if (codec->r_cfg.data)
 		rfree(codec->r_cfg.data);
 
+	codec->s_cfg.avail = false;
```

Comment:
@cujomalainey, is that this serious thing?

Reply:
Yes, looks like a memory leak

Reply:
ack, it's a leak. test is always false and we never free

Reply:
@lgirdwood  it's partially true - yes the logic here is flawed and the memory is not freed immediately after prepare, however, it will be reclaimed under .reset() and or .free() thanks to unconditional `rfree(codec->r_cfg.data);` present in both of these methods.

Reply:
> @lgirdwood it's partially true - yes the logic here is flawed and the memory is not freed immediately after prepare, however, it will be reclaimed under .reset() and or .free() thanks to unconditional `rfree(codec->r_cfg.data);` present in both of these methods.

Won't overwriting with NULL without freeing kinda stop that from working?

Reply:
ahh right! Sorry I got blind for a moment. I wonder how it passed our mem leak testcases though

Reply:
Do the test cases stress the setting of configurations while walking the pipeline state machine?

Reply:
Yes, in fact it doesn't need to stress that, even a single cycle should reveal that right? Need to review that test...

Reply:
I think as long as you apply runtime config then call prepare you should see the leak, unless we are still holding a reference in the layer above?

Reply:
I also thought there was other reference, that's why I suggested .free() and .reset() would reclaim that eventually, but it seems there is no other reference to that memory beside `rfree(codec->r_cfg.data);` So its good you found that.